packages/dcos-integration-test/extra/conftest.py

```diff
@@ -42,3 +49,45 @@ def dcos_api_session():
 @pytest.fixture(scope='session')
 def noauth_api_session(dcos_api_session):
     return dcos_api_session.get_user_session(None)
+
+
+@pytest.fixture(scope='session', autouse=True)
+def _dump_diagnostics(request, dcos_api_session):
+    """Download the zipped diagnostics bundle report from each master in the cluster to the home directory. This should
+    be run last. The _ prefix makes sure that pytest calls this first out of the autouse session scope fixtures, which
+    means that its post-yield code will be executed last.
+
+    * There is no official way to ensure fixtures are called in a certain order
+    https://github.com/pytest-dev/pytest/issues/1216
+    * However it seems that fixtures at the same scope are called alphabetically
+    https://stackoverflow.com/a/28593102/1436300
```

Comment:
Thanks for this docstring. It looks like we're relying on an undocumented implementation detail of pytest to make sure this fixture gets torn down last, which seems risky. Since pytest doesn't define the ordering of teardowns, meaning it could change, it seems like our best bet for the long term is to handle this outside of pytest. Were there any other options you explored? Perhaps we could implement this as a separate script (or a DC/OS CLI command?) that TC jobs explicitly run after tests. That would give us more confidence that the bundle covers the entire test run. And perhaps we could optimize space a little by limiting bundle collection to failed test runs.

Reply:
The long term plan is to get this functionality into https://github.com/dcos/dcos-test-utils and so that it could be used in `dcos-launch` directly, but I had originally hoped to get this out in time to be helpful for 1.11 debugging and reusing the code from the diagnostics tests was the simplest/quickest way.

For the dcos-integration-test pytest-based implementation I also looked into hooks that could be run at the end of a session, but if something isn't a fixture or a test, then it wouldn't be able to use the `dcos_api_session` fixture properly, and at that point I think it would be better to just go forward with the `dcos-test-utils`-based implementation.

The same effect can be done easily through the CLI with the `dcos node diagnostics` [`create`](https://docs.mesosphere.com/1.11/cli/command-reference/dcos-node/dcos-node-diagnostics-create/) and [`download`](https://docs.mesosphere.com/1.11/cli/command-reference/dcos-node/dcos-node-diagnostics-download/) commands (I think these ultimately call the same endpoints used here) which should be called with the [CLI module](https://github.com/dcos/dcos-test-utils/blob/master/dcos_test_utils/dcos_cli.py) in dcos-test-utils. I would have just put in these lines on the TC build script, but we currently don't install and set up the CLI on each integration test CI jobs because it's a lot of overhead.

re: artifact size: I can change that to only copy them over when tests fail.

Reply:
I'm currently looking into some of the ee diagnostics integration tests failing on the strict mode CI job. If it looks like those require any more finagling on this end I will just put this PR on hold and try doing it through dcos-test-utils anyway 😞

Reply:
Sounds good, I'm OK with merging this as a short-term solution (assuming you can get it working in strict security mode) and following up on it later.